lib/index.js

```diff
@@ -0,0 +1,211 @@
+    recoveryElectors,
+    // number of generations of merge events where no new electors participate
+    // and fewer than 2f+1 total electors (including self) participate causing
+    // recovery mode to trigger because consensus cannot be reached
+    recoveryGenerationThreshold,
+  };
+};
+
+// it is useful to override this function in tests
+api._computeElectors = async ({
+  blockHeight, electorPoolElectors, latestBlockSummary, ledgerConfiguration,
+  ledgerNode
+}) => {
+  // get partipicants for the last block
+  const {_blocks} = ledgerNode.consensus;
+  const {consensusProofPeers, mergeEventPeers} = await _blocks.getParticipants(
+    {blockHeight: blockHeight - 1, ledgerNode});
+
+  // TODO: we should be able to easily remove previously detected
+  // byzantine nodes (e.g. those that forked at least) from the electors
+
+  // the genesis node functions as a dictator until the electorPool document
+  // is established
+  if(electorPoolElectors.length === 0) {
+    return {electors: [{id: consensusProofPeers[0]}], recoveryElectors: []};
+  }
+
+  const {blockHash} = latestBlockSummary.eventBlock.meta;
+  const baseHashBuffer = Buffer.from(blockHash);
+  // the hash of the previous block is combined with the elector id to
+  // prevent any elector from *always* being sorted to the top
+  electorPoolElectors.sort((a, b) => {
+    // generate and cache hashes
+    a._hashBuffer = a._hashBuffer || xor(baseHashBuffer, Buffer.from(a.id));
+    b._hashBuffer = b._hashBuffer || xor(baseHashBuffer, Buffer.from(b.id));
+
+    // sort by hash
+    return Buffer.compare(a._hashBuffer, b._hashBuffer);
+  });
+
+  // TODO: allow the electorPool document to specify an electorCount. Use the
+  // value from the electorPool document first if available and if absent,
+  // use the value specified in the ledgerConfiguration.
+
+  let electorCount = Math.min(
+    ledgerConfiguration.electorSelectionMethod.electorCount,
+    electorPoolElectors.length);
+
+  // adjust electorCount to the form 3f+1
+  electorCount = _computeTargetElectorCount(
+    {coefficient: 3, originalCount: electorCount});
+
+  const recoveryElectorCount = _computeTargetElectorCount(
+    {coefficient: 1, originalCount: electorCount});
+
+  const electors = [];
+  const recoveryElectors = [];
+  const recoveryElectorCandidates = [];
+  const guarantorElectorCandidates = [];
+  const regularElectorCandidates = [];
+
+  // sort electorPoolElectors into categories based on elector.type
+  let i = electorPoolElectors.length;
+  while(i--) {
+    const elector = electorPoolElectors.pop();
+    const electorType = [].concat(elector.type);
+    if(electorType.includes(ELECTOR_TYPE_RECOVERY)) {
+      recoveryElectorCandidates.push(elector);
+    } else if(electorType.includes(ELECTOR_TYPE_GUARANTOR)) {
+      guarantorElectorCandidates.push(elector);
+    } else if(electorType.includes(ELECTOR_TYPE_REGULAR)) {
+      regularElectorCandidates.push(elector);
+    }
+  }
```

Comment:
We should add a TODO that will help us skip sorting this for every block via a cache and a sequence number on the elector pool record (or something like that). There could be thousands of electors in the pool.

Reply:
Also, the ordering needs to be deterministic or different nodes may pick different electors and BOOM.

Reply:
electorPoolElectors has already been sorted and therefore the category arrays will be sorted as well right?

Reply:
Oh, github hid that from me... it's already sorted, so nevermind the sort comments.